### ISSUE TITLE
Notifications/Badge counter Linux

### ISSUE THREAD

**Issue by exander77** [jiahaog/nativefier#962]
This is a discussion topic about how to tackle this.

It does not seem that icons support adding a number by default, so the idea is to embed the number in the icon. But generating icons on the fly seems complicated. So maybe pre generating icons with numbers and add them to application bundle?

Something like this:
![notify4](https://user-images.githubusercontent.com/2256039/80395075-03a6ad80-88b3-11ea-9032-6712592a6617.png)

**Comment by ronjouch** [comment]
> But generating icons on the fly seems complicated. So maybe pre generating icons with numbers and add them to application bundle?

@exander77 I'm not a fan of the idea of tons of repeated mini pngs in the bundle. How many of them? 10? 100? 1000? I know people with tens of thousands unread messages. Also, updating those will be a pain.

**Comment by exander77** [comment]
@ronjouch I am not a fan of this myself. :/

What I did so far for myself is 1, 2, 3, 4, 5, 6, 7, 8, 9 and + for more, I count new emails. If anybody has more than 9, it is probably hopeless to use such a counter at all.

This could also be different for notification counter instead of badge counter. As the notification counter is reset when the application has focus.

**Comment by ronjouch** [comment]
> What I did so far for myself is 1, 2, 3, 4, 5, 6, 7, 8, 9 and + for more, I count new emails. If anybody has more than 9, it is probably hopeless to use such a counter at all.

True, it's just a teeny small icon, especially compared to a macOS dock icon using `--counter` which has space to clearly display for example `193`. So your `1, 2, 3, 4, 5, 6, 7, 8, 9 and + for more` makes sense. Then, two thoughts:

1. I thought of what the Slack app does: just a bright red dot on unread messages (see screenshot below):
![screenshot_slack-unread-badge_2020-04-27-13:01:29](https://user-images.githubusercontent.com/522085/80399649-916ca380-8887-11ea-8fa0-fdd9f974728e.png)
2. I have one doubt about the screenshot you posted: some users run lots of nativefier apps, all with tray icons. So, a counter should not completely hide the icon (else you won't be able to differentiate them). See the slack icon above: the dot takes only one quarter of the icon.

So, given 1 + 2, how readable is the counter if taking only 1/4 (maybe a little bit more, letting you experiment) of the icon? If it's not readable, then maybe SlackApp's approach is the best we can do given the constraints

**Comment by exander77** [comment]
Telegram on Linux is actually showing the number of all messages, but it is very tiny.
![Screenshot_2020-04-27_00-02-33](https://user-images.githubusercontent.com/2256039/80400317-e83fa080-88ba-11ea-810e-b74a196ff96c.png)

1. The bright red dot is not a bad idea, I have seen that myself in many applications.
2. Yes, that is not a good example, I made it like this for myself, because of 22px bug I have been encountering. I think it should have a reasonable size.

**Comment by exander77** [comment]
This is my current state:
![icon](https://user-images.githubusercontent.com/2256039/80532618-b56dd900-899c-11ea-9fbe-f9ebb2d53be8.gif)
Feel free to comment.

**Comment by renyhp** [comment]
My two cents: I'd love to see this implemented, but I'd also love to see the icon(s) customizable, even if externally (e.g. they are stored somewhere I can replace with the icon I want).

**Comment by exander77** [comment]
@renyhp I have done it for myself, I had absolutely no time to share and I don't even know if such change would be accepted. I have basically a script that creates icons with badges based on svg template and target icon. And then I have a patch in nativefier to use it. I can share it with you and maybe we can put together some pull request.

**Comment by exander77** [comment]
Example:

![icon-notify-1](https://user-images.githubusercontent.com/2256039/95729586-63382000-0c7d-11eb-8936-d51b73bc03ac.png)
![icon-notify-10](https://user-images.githubusercontent.com/2256039/95729607-66cba700-0c7d-11eb-8f6b-94f92ea99b29.png)

**Comment by renyhp** [comment]
I know absolutely nothing of typescript, but yes, please, do share! If we can manage to make something usable and set it as an option we could most definitely file a PR.

**Comment by exander77** [comment]
Patched nativefier: https://github.com/exander77/nativefier
Scrip to generate icons: https://gist.github.com/exander77/1f0df9d6bdacae7daed1442217351927#file-build-icons-sh
You run script like this:
```
../build-icons.sh "YourApp-linux-x64/resources/app/icon.png"
```

You can try, report back for help and ideas.

**Comment by renyhp** [comment]
I have fiddled with your patch and with #967, but I have a problem with both. I can make it work when the window is minimized, but it seems that, if the window is closed, I do get notifications, but the tray icon is not updated...

**Comment by exander77** [comment]
@renyhp Interesting, what UI do you use? I am on Ubuntu 20.04 as well, with Xfce.

**Comment by exander77** [comment]
@renyhp Do you use switch `--counter`?

**Comment by renyhp** [comment]
That was it, thank you so much! I now have this other issue where if I set the `--name` switch, the `--tray start-in-tray` option has no effect - the app does stay in the tray, but it does not start in tray.
(Sorry if this is becoming a bit OT)

**Comment by renyhp** [comment]
I have looked through the code, but there are so many steps into the building of the icons and their paths, so I can't understand how to do anything other than simple copy-pasting.

I made a patch, which I've just realized is more relevant to #967 (I didn't implement an actual badge counter, which I personally don't need so much, although it would surely be a nice option to have) so I'll move my comments over there.

The issue I was mentioning in the previous comment seems to be WhatsApp specific and also present in the official release, so I'll maybe file a bug report.

**Comment by thamerx** [comment]
@exander77 

I was able to get it working on Windows. tho I had to modify a couple of things to get it working properly. 

I also added a badge to the taskbar icon not just for the tray icon. 

![Screenshot 2020-12-06 074316](https://user-images.githubusercontent.com/9639436/101271751-bcf92a00-3796-11eb-8ecb-e15514473a3d.jpg)

**Comment by exander77** [comment]
@thamerx I will look into it and maybe we can merge it so it works on both Linux and Windows.

**Comment by jacopo-j** [comment]
Hello, are there still plans to merge @PAHXO's edits?

**Comment by dagmoller** [comment]
Using canvas may help...
sample code:


```
    ...
    if (options.counter) {
      mainWindow.on('page-title-updated', (event, title) => {
        const counterValue = getCounterValue(title);
        if (counterValue) {
          appIcon.setToolTip(
            `(${counterValue})  ${options.name ?? 'Nativefier'}`,
          );
        } else {
          appIcon.setToolTip(options.name ?? '');
        }
        
        // change icon dynamically
        if (counterValue) {
            loadImage("icon.png").then(icon => {
                const canvas = createCanvas(icon.width, icon.height);
                const ctx = canvas.getContext('2d');
                ctx.drawImage(icon, 0, 0, icon.width, icon.height);
                
                var centerX = canvas.width / 2;
                var centerY = canvas.height / 2;
                var radius = 100;
                
                ctx.beginPath();
                ctx.arc(centerX, centerY, radius, 0, 2 * Math.PI, false);
                ctx.fillStyle = 'red';
                ctx.fill();
                ctx.lineWidth = 5;
                ctx.strokeStyle = 'black';
                ctx.stroke();
                
                ctx.font = '165px Arial'
                ctx.fillStyle = 'black';
                ctx.fillText(counterValue, (centerX / 2) - (value >= 10 ? 90 : 45), (centerY / 2) + 60);
                
                appIcon.setImage(nativeImage.createFromBuffer(canvas.toBuffer()));
            });
        } else {
            appIcon.setImage(nimage);
        }
      });
    } else {
    ...
```

**Comment by wokawoka** [comment]
any update about the badge notification count on linux?
it would just be great to have it mplemented

**Comment by exander77** [comment]
> any update about the badge notification count on linux? it would just be great to have it mplemented

There are several solutions suggested in this thread.

**Comment by wokawoka** [comment]
> > any update about the badge notification count on linux? it would just be great to have it mplemented
> 
> There are several solutions suggested in this thread.
then why is this issue still open?

unfortunately I can just see solutions implemented in some specific branches that have never been merged here
I was just asking if there are plans to finally merge them

**Comment by Folaht** [comment]
I don't want the counter, but do want the badge. Is there anything like that?
